qualtran/bloqs/mod_arithmetic/mod_subtraction.py

```diff
@@ -0,0 +1,230 @@
+        p: The modulus for the negation.
+
+    Registers:
+        x: The register contraining the integer we negate.
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585)
+        Fig 6b and 8
+    """
+
+    dtype: Union[QUInt, QMontgomeryUInt]
+    mod: 'SymbolicInt'
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('x', self.dtype)])
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', x: Soquet) -> Dict[str, 'SoquetT']:
+        if not isinstance(self.dtype.bitsize, int):
+            raise ValueError(f'symbolic decomposition is not supported for {self}')
+
+        ancilla = bb.allocate(1)
+        ancilla = bb.add(XGate(), q=ancilla)
+
+        x_arr = bb.split(x)
+        x_arr, ancilla = bb.add(
+            MultiControlX(cvs=[0] * self.dtype.bitsize), controls=x_arr, target=ancilla
+        )
+        x = bb.join(x_arr)
+
+        ancilla, x = bb.add(MultiTargetCNOT(self.dtype.bitsize), control=ancilla, targets=x)
+        (ancilla,), x = bb.add(
+            AddK(self.dtype.bitsize, self.mod + 1, cvs=(1,), signed=False), ctrls=(ancilla,), x=x
+        )
+
+        x_arr = bb.split(x)
+        x_arr, ancilla = bb.add(
+            MultiControlX(cvs=[0] * self.dtype.bitsize).adjoint(), controls=x_arr, target=ancilla
+        )
+        x = bb.join(x_arr)
+
+        ancilla = bb.add(XGate(), q=ancilla)
+        bb.free(ancilla)
+        return {'x': x}
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        cvs: Union[list[int], HasLength]
+        if isinstance(self.dtype.bitsize, int):
+            cvs = [0] * self.dtype.bitsize
+        else:
+            cvs = HasLength(self.dtype.bitsize)
+        return {
+            (MultiControlX(cvs), 1),
+            (MultiControlX(cvs).adjoint(), 1),
+            (MultiTargetCNOT(self.dtype.bitsize), 1),
+            (AddK(self.dtype.bitsize, k=self.mod + 1, cvs=(1), signed=False), 1),
+            (XGate(), 2),
+        }
+
+    def wire_symbol(
+        self, reg: Optional['Register'], idx: Tuple[int, ...] = tuple()
+    ) -> 'WireSymbol':
+        if reg is None:
+            return Text("")
+        if reg.name == 'x':
+            return TextBox('$-x$')
+        raise ValueError(f'Unrecognized register name {reg.name}')
+
+
+@frozen
+class CModNeg(Bloq):
+    r"""Performs controlled modular negation.
+
+    Applies the operation $\ket{c}\ket{x} \rightarrow \ket{c}\ket{(-1)^c x\%p}$
+
+    Args:
+        dtype: Datatype of the register.
+        p: The modulus for the negation.
+        cv: value at which the gate is active.
+
+    Registers:
+        ctrl: Control bit.
+        x: The register contraining the integer we negate.
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585)
+        Fig 6b and 8
```

Comment:
if this doesn't match the reference, please provide a short description of how/why in the class docstring

Reply:
added